syz-manager/covfilter.go

```diff
@@ -52,15 +52,15 @@ func (mgr *Manager) createCoverageFilter() ([]byte, map[uint32]uint32, error) {
 	if !mgr.cfg.SysTarget.ExecutorUsesShmem {
 		return nil, nil, fmt.Errorf("coverage filter is only supported for targets that use shmem")
 	}
-	bitmap := createCoverageBitmap(mgr.cfg.SysTarget, pcs)
+	bitmapPCs := pcs
```

Comment:
Thanks for the catch; that's a good point. I've updated it so the map is copied.

I'm not sure if both are needed? According to this comment (https://github.com/google/syzkaller/blob/master/syz-manager/covfilter.go#L56-L62), the only difference is that pcs from symbol CMPs are not included.

I did test adding symbol CMP functions just before creating the bitmap, but the addition still needs to be saved in the manager (or server) to avoid re-creating the same bitmap PCs for each instance. However, so that fewer bitmaps are passed back I could implement this.

As a side note, the comment I referenced implies that all CMP pcs should not be included. Should this be extended for unit CMPs as well?

Reply:
I'd check whether excluding cmp PCs from the coverage bitmap could affect comparison values collection in the executor. If (most likely) it doesn't affect it at all, I'd go with removing those PCs and returning just one map from `createCoverageFilter()`. In the end, why filter with one set of PCs on VMs and with another (smaller one) in `syz-manager`..

> Should this be extended for unit CMPs as well?

I don't have a definite answer. Are there many unit cmp PCs that are also not symbol cmp PCs?

Reply:
> I don't have a definite answer. Are there many unit cmp PCs that are also not symbol cmp PCs?

Good point; no, in my tests all unit CMP PCs were also symbol cmp PCs.

> In the end, why filter with one set of PCs on VMs and with another (smaller one) in syz-manager.

It looks like the syz-manager doesn't actually filter the corpus; it just adds coverage values which do fall into the smaller filter to the `filtered coverage` statistic (https://github.com/google/syzkaller/blob/master/syz-manager/rpc.go#L304).

Looking at the manager of `ci2-upstream-f2`, which does use filtering, the `coverage` vs `filtered coverage` do have a pretty significant difference in values (http://34.123.30.149:50026/), so I'm hesitant about removing this completely.

Reply:
Ah, interesting. So we need that map only to calculate one stats value. If I'm not mistaken and if CMP and non-CMP PCs don't intersect, we really need it just for `len(mgr.coverFilter)` from here:

https://github.com/google/syzkaller/blob/4cd5bb25a2752a9a5b25597d1da34656681f07a6/syz-manager/html.go#L143

Maybe, instead of returning two maps, just return a single map and an integer (# of filtered PCs) then? This should hopefully make the code less confusing.

> I'd check whether excluding cmp PCs from the coverage bitmap could affect comparison values collection in the executor.

It does affect:
https://github.com/google/syzkaller/blob/4cd5bb25a2752a9a5b25597d1da34656681f07a6/executor/executor.cc#L1615-L1617

Though I'm not sure whether it's good or bad. When we mutate with hints, we'll in the end anyway look at the (new) signal. And that signal is already filtered.

If we don't filter kcov cmp output at all, on the one hand, we'll get comparison arguments from the code that is filtered out and those values will likely be useless for us. But, on the other hand, it could well be the case that the target (filtered) kernel code is reached via non-filtered code, so it could lead to better coverage of the filtered code..

Reply:
I missed it before, but it looks like it's also used when filtering PCs for the corpus in pkg/cover/html.go:

https://github.com/google/syzkaller/blob/4cd5bb25a2752a9a5b25597d1da34656681f07a6/pkg/cover/html.go#L570-L572 

I do like the idea of only needing an integer, but if we need to filter out the PCs for the corpus the full map may be necessary. We could also include CMP values in all statistics on the syz-manager side, though that would change the filtered coverage values.

Another option could be to pass back the array of sym.CMPs: https://github.com/google/syzkaller/blob/4cd5bb25a2752a9a5b25597d1da34656681f07a6/syz-manager/covfilter.go#L58-L61

This would still require two values to be returned, but as executor values are already filtered, only the CMP values would be needed to filter out values in the manager which may make the filter function more clear as well.

I'm currently setting up a syz-testbed instance to check how execution changes with no filtered CMP PCs in the executor.

Reply:
It's still early/has a low sample size, but it looks like the number of crashes has dropped without filtered CMP PCs in the executor. The test branch I used is here (https://github.com/google/syzkaller/compare/master...eprucka3:syzkaller:filter_cmp_test), and only differs in that the CMPs have been deleted before creating the coverage bitmap.
![image](https://github.com/google/syzkaller/assets/33524786/c8d4bebf-b4b5-454a-905b-6d68b66dfe70)

Reply:
Yes, it's way too early yet to draw any conclusions. I'd trust results if there are at least 10-15 runs for each branch. It sometimes helps to set it up on GCE to speed up things (it will need a lot of resources, but at least there'll be results in a short time).

Reply:
I think it's the opposite; we return cmp and non-cmp PCs from syz-executor/syz-fuzzer, but `coverfilter` only uses non-cmp PCs

https://github.com/google/syzkaller/blob/4cd5bb25a2752a9a5b25597d1da34656681f07a6/syz-manager/covfilter.go#L56-L62

For context, `bitmap` is sent to the executor, and `pcs` is used by syz-manager as `coverfilter`.

< Yes, it's way too early yet to draw any conclusions. I'd trust results if there are at least 10-15 runs for each branch. It sometimes helps to set it up on GCE to speed up things (it will need a lot of resources, but at least there'll be results in a short time).

Sounds good, I'll set this up.

Reply:
Looked once more into the code. 

`syz-executor` filters signal, but returns coverage as is. It's rather unlikely that it contains CMP coverage -- we either collect comparison arguments or normal coverage. `syz-manager` then collects both the unfiltered coverage it received with each input and, separately, filtered coverage:

https://github.com/google/syzkaller/blob/1a2f6297df2e11f3ef37e97803568cb1b9ef875b/syz-manager/rpc.go#L292-L304

Thus the difference between `coverage` and `filtered coverage` stats.

So let's just return two maps, as we already do now -- one for `cmp+pc`, one for `pc` only.  Just add comments/rename variables so that it's easier to see the motivation and the difference between those maps.

Reply:
Sounds good! Done.